### ISSUE TITLE
[feat] Build for the web

### ISSUE THREAD

**Issue by Jamzy01** [tauri-apps/tauri#8248]
### Describe the problem

I know that Yew already exists, but implementing a way to build for the web in tauri would be a really useful feature

### Describe the solution you'd like

I don't need this feature urgently, but it would be nice if it could at least be added to the Tauri roadmap by an admin to be implemented at some point. I just want a simple way to build a single page app (or even multi page at some point). I think it'd be a great idea because then the web app can leverage the power of rust whilst still being easy to write and having nice gui.

### Alternatives considered

Of course, there's yew, but that'd mean you'd have to create your app again from the ground up.

**Comment by amrbashir** [comment]
I am not sure what you mean tbh so would need more clarification. Tauri uses web technology to present the UI inside a webview on Desktop, you can still reuse your HTML/CSS/JS project and run it inside the browser (you still need to guard your usage of tauri APIs though so it doesn't error in the browser)

**Comment by Jamzy01** [comment]
> I am not sure what you mean tbh so would need more clarification. Tauri uses web technology to present the UI inside a webview on Desktop, you can still reuse your HTML/CSS/JS project and run it inside the browser (you still need to guard your usage of tauri APIs though so it doesn't error in the browser) 

Yeah I know I wasn't particularly clear, and the reason for that is that I'm not 100% sure of how I'd like it to be implemented. Now yes, I know I can just run it inside a web browser, but as you said, then you can't call tauri commands, and you just overall don't have the rust backend. And I know that for my app, my frontend utilizes the backend for it's speed to perform heavy computations even when native apis aren't even being used (in other words, I'm writing code in rust that could just as well be written in js, but I'm doing it to increase performance). And allowing normal websites to integrate with a rust backend (a backend that I assume would be ran though wasm) would be great because then you can make faster websites still use the same web technologies you're used to (unlike yew where you have to mess around learning their own web framework, which makes it difficult to port from an existing web project).

Also something that would be nice would be some inbuilt apis to rust for the web, like for example still allowing the backend to make http requests, still giving dialog support etc).

And of course I'd want a way to conditionally compile code only for a web build, something like this:

`#[cfg(web)]`

The current app I'm working on should work great as a browser build, because with a few small tweaks it'd work just fine without native apis. And my app would definitely be much faster than just a regular react app levering the power of rust. And I'm also sure that if apps like discord and teams were made using tauri, then this'd make it easy for them to have one codebase for all platforms.

Honestly I'm not sure if this clears up how it should be implemented if it were to be, but I hope it explains why I want it implemented and what its use case could be.

**Comment by amrbashir** [comment]
Thanks for the clear up, I understand the motivation and while It is a valid case I don't think we have enough resources to spend on this at the moment so I will keep this issue open in case we get some time for it or if someone else wants to pick it up.

**Comment by Jamzy01** [comment]
I know just how incredibly hard of a feature it would be to implement, but I was just wondering if it could be added to the tauri roadmap as a very low priority feature that could be implemented at some point if someone feels like doing it.

**Comment by tetuaoro** [comment]
Hello, tauri can be integrated into an existing web project. All you need to do is define a few parameters, such as the location of the output folder, and you're ready to go. However, I find your idea interesting.

**Comment by Tnze** [comment]
@Jamzy01 I have a working project running both on Desktop and Web.
Here: <https://github.com/Tnze/ffxiv-best-craft>

So basically I have a `lib` crate for my rust code. And a tauri binding for desktop and wasm binding for web application.

Having two head-writing binding is a bit annoying. But at least it works.

**Comment by Jamzy01** [comment]
> @Jamzy01 I have a working project running both on Desktop and Web. Here: https://github.com/Tnze/ffxiv-best-craft
> 
> So basically I have a `lib` crate for my rust code. And a tauri binding for desktop and wasm binding for web application.
> 
> Having two head-writing binding is a bit annoying. But at least it works.

Your game looks really great! It's a proof of concept that something like this will work just fine, but it'd just be so much nicer if it was built into Tauri.

**Comment by Tnze** [comment]
> It'd just be so much nicer if it was built into Tauri.

I think we can build a new wasm binding tools that easy to used just like Tauri's #[tauri::command]. Then we can write a crate that can switch between compiled to a WASM binary or a Tauri plugin.

**Comment by AndrewJSchoen** [comment]
Just my two cents: I have talked with _many_ people that are looking to switch to a different app development platform, and who were curious about my thoughts, since I have used Tauri on numerous occasions. I told them that Tauri was great for building entirely self-contained web-apps in an app wrapper (e.g. effectively no rust backend), and great for dedicated desktop apps that rely on a rust backend. I have used it previously for both, and I can attest to that. However, when I tell them that it is not great for web+desktop with a mostly shared codebase - precisely because of the fact that it ends up being completely on the developer to architect those switches and builds - it turns out that it is a huge red flag for people that turns them away from the project. 

As I have gotten more familiar with Rust, it feels odd that there aren't feature flags/configuration conditionals that allow users to define both web implementations and native ones for when the available libraries differ, and then support building the app in either web or desktop/mobile through a unified build. I understand that the tauri devs are concerned about the lack of support for some desktop functionality in wasm-based builds, but it seems like there is a lot of other configuration that ends up being needed to figure out a web version that ends up being put on the app developer, but which actually could be handled easily in the build process with sufficient build-time conditionals (e.g. wasm build, function linking, worker execution, etc).

**Comment by ecstrema** [comment]
I have the exact same experience as @AndrewJSchoen above. I write some functionality in rust so performance critical code stays native, but I still like to have a web build, so I ended building a js wrapper against the wasm-pack version and the tauri version. It looks approximately like this:

```ts
// wrapper.ts

import my_wasm_lib from 'my_wasm_lib';


export function foo(bar: string) {
  if (IS_NATIVE) {
    invoke('foo', { bar });
  else {
    my_wasm_lib.foo(bar);
  }
}
```

Then, rust commands look like this:

```rust
#[wasm_bindgen]
#[tauri::command]
fn foo(s: String) {
  println!("I was invoked from JavaScript, with this message: {}", invoke_message);
}
```

However, this is tedious, repetitive work, so I think that:

- `#[tauri::command]` should also add the attribute`#[wasm_bindgen]`
- An additional file should be added to the wasm_bindgen output containing the wrapper.ts like above
- the package.json file generated by wasm bindgen should point to it instead.

Since `IS_NATIVE` is defined in my case by vite so that it's optimized out by the build tool, I don't know how that can be replaced by tauri directly without relying on vite...

**Comment by Eveeifyeve** [comment]
I feel like this would be better for v3 in tauri as it will consider a lot of a work and will require a major change.

**Comment by FabianLars** [comment]
more like v4 or v5 if ever - v3 won't get that many (major) changes.

**Comment by keadex** [comment]
Not sure if this will be helpful, but here’s how I built a web version of a Tauri desktop app (Keadex Mina) - while minimizing the need for `if` conditions or changes to the original Tauri code.

[Keadex Mina](https://keadex.dev/mina) is an open-source, cross-platform application I built with Tauri. Its codebase is about **70% Rust** (backend & business logic) and **30% TypeScript/React** (frontend/UI).

Originally designed as a desktop app (Windows, Linux, macOS), I later ported it to the web:

- **Desktop:** [Release v2.8.0](https://github.com/keadex/keadex/releases/tag/keadex-mina-v2.8.0)  
- **Web:** [Mina Live](https://keadex.dev/en/mina-live)


### Key architectural decisions

- **New library – Mina Live**: Reuses ~95% of the Rust and TypeScript code from the Tauri app. Rust is compiled to Wasm, and React imports it.  
  - [Source code](https://github.com/keadex/keadex/tree/main/libs/mina-live)  
  - [NPM package](https://www.npmjs.com/package/@keadex/mina-live)

- **Remaining 5%** of Mina Live code covers high-level modules:  
  - **Tauri adapters** – let me port the Tauri ecosystem to the web and route Rust calls to Wasm without changing app code:  
    - [Tauri Web Adapter](https://github.com/keadex/keadex/blob/main/libs/mina-live/src/tauri/tauri-web-adapter.ts) – mocks runtime to automatically invokes web-based plugins or wasm code
    - [Tauri Plugin Adapter](https://github.com/keadex/keadex/blob/main/libs/mina-live/src/tauri/tauri-plugin-adapter.ts) – mocks plugins 
    - [Wasm Bridge](https://github.com/keadex/keadex/blob/main/libs/mina-live/src/tauri/wasm-bridge.ts) – intercepts Rust calls from web and invokes Wasm  
  - **Webpack configs** – override Tauri API imports with adapter versions ([example](https://github.com/keadex/keadex/blob/d3a619da19c2dea7f3ad91f97eb9b87af01493ac/libs/mina-live/webpack.config.ts#L86))  
  - **Mina Live React component** – the main entrypoint exported by the library, embeddable in other React apps ([code](https://github.com/keadex/keadex/blob/main/libs/mina-live/src/components/MinaLive/MinaLive.tsx)).
